[SPARK-33888][SQL] JDBC SQL TIME type represents incorrectly as TimestampType, it should be physical Int in millis

saikocat opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
JDBC SQL TIME type represents incorrectly as TimestampType, we change it to be physical Int in millis for now.

<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
Currently, for JDBC, SQL TIME type represents incorrectly as Spark TimestampType. This should be represent as physical int in millis Represents a time of day, with no reference to a particular calendar, time zone or date, with a precision of one millisecond. It stores the number of milliseconds after midnight, 00:00:00.000.


<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce _any_ user-facing change?
No
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->


### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->
Close #30902

saikocat committed:
[SPARK-33888] Add support for TimeMillis logicalType to TimestampType

AmplabJenkins commented:
Can one of the admins verify this patch?

gengliangwang commented:
Hi @saikocat ,

thanks for reporting the issue. 
As per Avro spec https://avro.apache.org/docs/1.8.2/spec.html#Time+%28millisecond+precision%29:
```
Time (microsecond precision)
The time-micros logical type represents a time of day, with no reference to a particular calendar, time zone or date, with a precision of one microsecond.

A time-micros logical type annotates an Avro long, where the long stores the number of microseconds after midnight, 00:00:00.000000.

Timestamp (millisecond precision)
The timestamp-millis logical type represents an instant on the global timeline, independent of a particular time zone or calendar, with a precision of one millisecond.

A timestamp-millis logical type annotates an Avro long, where the long stores the number of milliseconds from the unix epoch, 1 January 1970 00:00:00.000 UTC.
```
So, we can't map the time-millis/time-micros type to Timestamp type. Instead, we should map it as `CalendarIntervalType`.
In Spark, `CalendarIntervalType` is internal type and we can't use it for table creatation. So we didn't support the Avro logical type time-millis and time-micros.
In the current stage, I am +0 for supporting it.
cc @cloud-fan

saikocat commented:
Hi gengliangwang,

Thank you so much for your reply. I understand the issue now. 

Then I guess the bug should be in the Spark JDBC Utils schema conversion part then? Cos it converts a `TIME(6)` in MySQL into `TimestampType`. It should be `IntegerType` then no?

gengliangwang commented:
>Then I guess the bug should be in the Spark JDBC Utils schema conversion part then (https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala#L229)? Cos it converts a TIME(6) in MySQL into TimestampType. It should be IntegerType then no?

Yes, it should be Integertype or CalendarIntervalType

cloud-fan commented:
TIME is a standard SQL data type, which is not TIMESTAMP or INTERVAL. Unfortunately, Spark doesn't support this data type yet, and can only fail (or read as the physical type int).

The JDBC mapping is wrong and we should fix it.

saikocat commented:
Noted with thanks. Should I just make a change/fix to `JdbcUtils.scala` then? Should be an easy fix if we choose to go with `IntegerType`.

cloud-fan commented:
Yea, and we can also fix avro to produce better error message (or read time as int)

hoand-teko committed:
[SPARK-33888] Fix JDBC Mapping for JDBC TIME type

saikocat committed:
[SPARK-33888] Fix JDBC Mapping for JDBC TIME type

saikocat committed:
[SPARK-33888] Fix unit test

saikocat committed:
[SPARK-33888] Fix unreachable code

saikocat committed:
[SPARK-33888] Address scalastyle check

saikocat committed:
[SPARK-33888] Update documentation for mapping

saikocat committed:
[SPARK-33888] Add unit test for TIME types

saikocat commented:
I think I've fixed the type mapping and tests. However, I can't seem to pass this test `CliSuite.SPARK-29022: Commands using SerDe provided in --hive.aux.jars.path ` which looks pretty much unrelated to the changes I made: https://github.com/apache/spark/pull/30902/checks?check_run_id=1607685366.

Any helps would be much appreciated!

Merry Christmas!

cloud-fan commented:
ok to test

SparkQA commented:
**[Test build #133380 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133380/testReport)** for PR 30902 at commit [`68e2afe`](https://github.com/apache/spark/commit/68e2afe6d6979e6240cbf1fd62cd1f02e870e0a8).

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/37971/

saikocat commented:
Seems like Amplab Jenkins CI run out of space when assembly :(

SparkQA commented:
**[Test build #133380 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133380/testReport)** for PR 30902 at commit [`68e2afe`](https://github.com/apache/spark/commit/68e2afe6d6979e6240cbf1fd62cd1f02e870e0a8).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/133380/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -409,8 +410,25 @@ object JdbcUtils extends Logging {
+        // Represents a time of day, with no reference to a particular calendar,
+        // time zone or date, with a precision of one millisecond.
+        // It stores the number of milliseconds after midnight, 00:00:00.000.
+        if (rs.getMetaData.getColumnType(columnIdx) == java.sql.Types.TIME) {
```
do we run the `if` check for every input row?

saikocat replied:
Yes. You're right.

Originally, I  made changes to private method `makeGetters` and pass in the `ResultSetMetadata` as additional parameter. Then we have a separate pattern matching for `case IntegerType if rsMetadata.getColumnType(columnIdx) == TIME =>`  But right now, the column index/position is not accessible at the pattern matching level, only available at the closure function :(

Any advice is deeply appreciated here.

saikocat replied:
Actually, let me see if I can work with the `metadata` from `getSchema` function.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -628,6 +635,20 @@ class JDBCSuite extends QueryTest
+    val expectedTimeMillis = java.util.concurrent.TimeUnit.SECONDS.toMillis(
+      expectedTimeRaw.toLocalTime().toSecondOfDay()
+    ).toInt
+    assert(rows(0).getAs[java.sql.Time](0) === expectedTimeMillis)
```
what was the result before? null?

saikocat replied:
Result before is `expectedTimeMilllis` for `rows(0)`.
For `row(1)` then it is `null`

cloud-fan commented:
@saikocat  can you update the PR description? It's not related to avro now.

saikocat committed:
[SPARK-33888] Fix metadata builder for getSchema()

saikocat committed:
[SPARK-33888] Use int if logicaltimetype metadata is present

saikocat commented:
@cloud-fan I have addressed the code review comments. Took awhile cos of a bug in metadata builder not being build.

I will use the metadata field to store a key `logicaltimetype` boolean and check for this in pattern matching.
Also fixed a bug in the metadata not being build/propagate for `StructField` in `getSchema()`.  Ideally another MR but this fix is required for the extra field in metadata.

SparkQA commented:
**[Test build #133446 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133446/testReport)** for PR 30902 at commit [`c75037a`](https://github.com/apache/spark/commit/c75037ad1e534b20463b3b86cda385f593796d94).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38035/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38035/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38035/

SparkQA commented:
**[Test build #133446 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133446/testReport)** for PR 30902 at commit [`c75037a`](https://github.com/apache/spark/commit/c75037ad1e534b20463b3b86cda385f593796d94).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/133446/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -408,6 +421,23 @@ object JdbcUtils extends Logging {
+      (rs: ResultSet, row: InternalRow, pos: Int) => {
+        val rawTime = rs.getTime(pos + 1)
+        if (rawTime != null) {
+          val rawTimeInSeconds = rawTime.toLocalTime().toSecondOfDay()
```
so the JDBC TIME only has second precision?

cloud-fan replied:
`toSecondOfDay` will truncate the value to second precision, and then we lose the millisecond precision. I think we should call `toNanoOfDay` and then truncate to millisecond.

saikocat replied:
Yup. Only till millis based on the constructor https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/Time.html#<init>(long) . But database like MySQL can store up till microseconds.

Thanks for the review. I've change to nano then truncate it. Also add an additional test for H2 time type as well.

saikocat commented:
Yup. Only till millis based on the constructor
https://docs.oracle.com/en/java/javase/11/docs/api/java.sql/java/sql/Time.html#<init>(long)
. But database like MySQL can store up till microseconds.

On Tue, 29 Dec 2020, 10:28 pm Wenchen Fan, <notifications@github.com> wrote:

> *@cloud-fan* commented on this pull request.
> ------------------------------
>
> In
> sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala
> <https://github.com/apache/spark/pull/30902#discussion_r549724512>:
>
> > @@ -408,6 +421,23 @@ object JdbcUtils extends Logging {
>        (rs: ResultSet, row: InternalRow, pos: Int) =>
>          row.setFloat(pos, rs.getFloat(pos + 1))
>
> +
> +    // SPARK-33888 - sql TIME type represents as physical int in millis
> +    // Represents a time of day, with no reference to a particular calendar,
> +    // time zone or date, with a precision of one millisecond.
> +    // It stores the number of milliseconds after midnight, 00:00:00.000.
> +    case IntegerType if metadata.contains("logicaltimetype") =>
> +      (rs: ResultSet, row: InternalRow, pos: Int) => {
> +        val rawTime = rs.getTime(pos + 1)
> +        if (rawTime != null) {
> +          val rawTimeInSeconds = rawTime.toLocalTime().toSecondOfDay()
>
> so the JDBC TIME only has second precision?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/spark/pull/30902#pullrequestreview-559556378>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AA62IDKK7C5SREGWBT3L3ULSXHRRVANCNFSM4VGUOSUQ>
> .
>

saikocat committed:
[SPARK-33888] Fix loss precision - use nano then truncate to millis

saikocat committed:
[SPARK-33888] Add test for H2 time type with milliseconds

SparkQA commented:
**[Test build #133503 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133503/testReport)** for PR 30902 at commit [`f55a7d6`](https://github.com/apache/spark/commit/f55a7d6eaab5f9514a0dabcdc83738ed21562139).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38092/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38092/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38092/

SparkQA commented:
**[Test build #133503 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133503/testReport)** for PR 30902 at commit [`f55a7d6`](https://github.com/apache/spark/commit/f55a7d6eaab5f9514a0dabcdc83738ed21562139).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/133503/

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -408,6 +421,23 @@ object JdbcUtils extends Logging {
+        val rawTime = rs.getTime(pos + 1)
+        if (rawTime != null) {
+          val rawTimeInNano = rawTime.toLocalTime().toNanoOfDay()
+          val timeInMillis = TimeUnit.NANOSECONDS.toMillis(rawTimeInNano).toInt
```
Can we use `Math.toIntExact` to avoid overflow?

saikocat replied:
:+1: Thanks for this! Good catch!

saikocat committed:
[SPARK-33888] Use Math.toIntExact to avoid overflow

SparkQA commented:
**[Test build #133519 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133519/testReport)** for PR 30902 at commit [`6d5a3c4`](https://github.com/apache/spark/commit/6d5a3c406102579bde264cfff29bc58dfd74773b).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38108/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38108/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38108/

SparkQA commented:
**[Test build #133519 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133519/testReport)** for PR 30902 at commit [`6d5a3c4`](https://github.com/apache/spark/commit/6d5a3c406102579bde264cfff29bc58dfd74773b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/133519/

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -625,9 +632,24 @@ class JDBCSuite extends QueryTest
     assert(rows(0).getAs[java.sql.Timestamp](2).getNanos === 543543000)
   }
 
+  test("SPARK-33888 : test TIME types") {
```
nit: `SPARK-33888: ...`

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -303,11 +304,23 @@ object JdbcUtils extends Logging {
+        // scalastyle:off
+        case java.sql.Types.NUMERIC => metadata.putLong("scale", fieldScale)
+        case java.sql.Types.DECIMAL => metadata.putLong("scale", fieldScale)
+        case java.sql.Types.TIME    => metadata.putBoolean("logicaltimetype", true)
```
how about `logical_time_type `

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/jdbc/JDBCSuite.scala:
```diff
@@ -625,9 +632,24 @@ class JDBCSuite extends QueryTest
+      expectedTimeRaw.toLocalTime().toSecondOfDay()
+    ).toInt
+    assert(rows(0).getAs[java.sql.Time](0) === expectedTimeMillis)
+    assert(rows(1).getAs[java.sql.Time](0) === expectedTimeMillis)
```
are we comparing `java.sql.Time` and int? How does it work?

saikocat replied:
Oops. The value is already converted to Int via `makeGetter` in JdbcUtils. So these should be `getAs[Int]`. Sorry about that.

saikocat committed:
[SPARK-33888] Address code review comment: snake_case, spacing, test

SparkQA commented:
**[Test build #133537 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133537/testReport)** for PR 30902 at commit [`63fad20`](https://github.com/apache/spark/commit/63fad20278c90b55fc311f276008ead5883dba1d).

SparkQA commented:
Kubernetes integration test starting
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38126/

SparkQA commented:
Kubernetes integration test status success
URL: https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder-K8s/38126/

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/38126/

SparkQA commented:
**[Test build #133537 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/133537/testReport)** for PR 30902 at commit [`63fad20`](https://github.com/apache/spark/commit/63fad20278c90b55fc311f276008ead5883dba1d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/133537/

saikocat commented:
Anything else that is needed before this can be merged?

AmplabJenkins commented:
Can one of the admins verify this patch?

cloud-fan commented:
thanks, merging to master!

dongjoon-hyun commented:
Hi, All.
This seems to break JDBC IT. Please see @sarutak 's PR.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -408,6 +421,23 @@ object JdbcUtils extends Logging {
+    // It stores the number of milliseconds after midnight, 00:00:00.000.
+    case IntegerType if metadata.contains("logical_time_type") =>
+      (rs: ResultSet, row: InternalRow, pos: Int) => {
+        val rawTime = rs.getTime(pos + 1)
```
@sarutak do you mean what returns here is seconds (with certain precision) from a starting timestamp, while the timestamp is different between databases? I'm a bit surprised if the JDBC protocol was design this way, but if this is true, then this PR doesn't make sense...

cloud-fan replied:
What returns here is `java.sql.Time`, and its doc says
```
The date components should be set to the "zero epoch"
value of January 1, 1970 and should not be accessed.
```
Maybe some databases don't follow the requirement, but it doesn't matter, as we call `rawTime.toLocalTime` which only access the hour:minute:second components.

cloud-fan replied:
That said, I think reading SQL TIME type as integer has a well-define semantic in Spark (after this PR): the integer represents the milliseconds of the time from 00:00:00.

sarutak replied:
O.K, so the test seems wrong.
Actually, I'll fix it in another PR.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/jdbc/JdbcUtils.scala:
```diff
@@ -408,6 +421,23 @@ object JdbcUtils extends Logging {
+
+    // SPARK-33888 - sql TIME type represents as physical int in millis
+    // Represents a time of day, with no reference to a particular calendar,
+    // time zone or date, with a precision of one millisecond.
```
After a second thought, why do we pick millisecond precision? Why not microsecond? Is there a standard for it?

saikocat replied:
Since we are converting to/from java.sql.Time, and according to the javadoc https://docs.oracle.com/javase/8/docs/api/java/sql/Time.html , it supports till milliseconds for constructor.

cloud-fan replied:
It may confuse Spark users, as Spark timestamp is microsecond precision.

After more thought, it's probably better to return timestamp when reading JDBC time, with a clear rule: we convert the time to timestamp by using "zero epoch" as the date part. It's also more useful as users can call `hour` function or similar ones to get some field values. What do you think?

saikocat replied:
Hmm, I agree with you on the user experience part (`hour` function and all). I think it is hard (near impossible) to introduce new DataType (Time - HH:MM:SS.sss display) and another function to convert an int without date portion to Timestamp (most conversion is parse string) as it gonna take through multiple level of approvals and testings.

Another reason why I went with Int was that with TimestampType, it breaks compatibility with Avro logical & Spark type converter (https://github.com/apache/spark/blob/master/external/avro/src/main/scala/org/apache/spark/sql/avro/SchemaConverters.scala).

I'm not sure if a `time_to_timestamp()` helper function would be a better compromise or revert back the 2 MRs?

cloud-fan replied:
It looks better if the avro schema converter can convert timestamp to time. After reading time column from JDBC, it becomes `IntegerType` and there is no context to indicate that this int comes from JDBC time and means milliseconds. What if the avro logic type is time-micros? With timestamp type, at least we know the precision is microsecond.

saikocat replied:
Your suggestion makes sense. We can also stuff the info into the metadata field of the struct field. Let me find sometimes this weekend or over the new year to try out your suggestion? Will ping you back if I couldn't find time or manage to get a solution so we can revert the MRs. Thanks a lots for helping me out!

Should I create a new JIRA ticket and cut a new MR or do you prefer to consolidate in here?

cloud-fan replied:
Let's create a new JIRA to track it. The PR is merged to master only so we have plenty of time to fix it before the 3.2 release :)

saikocat replied:
@cloud-fan btw, I did a quick check and it seems like if we use TimestampType, the time will always be converted to the JVM system timezone. So for the JDBCSuite test, given `12:34:56` time value, when you do the `.select(hour("time"))` it will always point to my local timezone hour instead of `12`. So I don't know if we should proceed in this case.

cloud-fan replied:
We can do whatever we want. We can use JDBC API `getTime` to get the time value, and construct the timestamp value in a reasonable way. It's under our control.
